src/js/CheckboxTree.js

```diff
@@ -212,64 +235,152 @@ class CheckboxTree extends React.Component {
             showNodeTitle,
             showNodeIcon,
         } = this.props;
-        const { id, model } = this.state;
-        const { icons: defaultIcons } = CheckboxTree.defaultProps;
 
+        const { icons, id } = this.state;
+
+        let state1counter = 0;
```

Comment:
These `statecounter` variables are very scary. I admittedly have little idea of what is going on here but I am averse to seeing variables of the format `var1` and `var2`. I would prefer more semantically named variables.

Reply:
These are semantically named variables.  :)  

`state1counter` is counting the number of nodes which have `checkstate === 1` in the current call of the recursive function `renderTreeNodes`.  `state2counter` does the same for  `checkstate === 2`.

I was thinking of using `checkstate1counter` or `checkstateEquals1Counter` but opted for the shorter versions. Would you prefer one of the longer names? If not, I should probably explain it in the comments.

As for your understanding what is going on, I probably need better comments.

Each call to `renderTreeNodes` is processing the children of a parent node. These counters are tabulating the number of children with 'full' or 'partial' checks, `checkstate` 1 and 2 respectively.

These counters are returned from `renderTreeNodes` as `numFullcheck` and `numPartialcheck`. These variables are then used in the calculation of the `checkstate` of the parent node. This parent node is in turn the child of another parent and thus tabulated by the counter for its parent.

Let me know if this does not make sense. It took me a while to figure out how to do this in an efficient fashion.